### PR TITLE
quic: add Private Key Provider support

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: quic: add Private Key Provider support
Additional Description:

This PR adds private key provider support to the QUIC over TLS 1.3

Risk Level: low
Testing: unittest and integration test
Docs Changes: api
Release Notes: new feature
Platform Specific Features: n/a

### CODE REVIEW

**Comment by soulxu** [comment]
cc @danzh2010 @ggreenway @yanavlasov

File: source/extensions/quic/crypto_stream/envoy_quic_crypto_server_stream.cc [hunk 1]
```diff
@@ -8,10 +11,12 @@ EnvoyQuicCryptoServerStreamFactoryImpl::createEnvoyQuicCryptoServerStream(
     const quic::QuicCryptoServerConfig* crypto_config,
     quic::QuicCompressedCertsCache* compressed_certs_cache, quic::QuicSession* session,
     quic::QuicCryptoServerStreamBase::Helper* helper,
-    // Though this extension doesn't use the two parameters below, they might be used by
-    // downstreams. Do not remove them.
-    OptRef<const Network::DownstreamTransportSocketFactory> /*transport_socket_factory*/,
-    Envoy::Event::Dispatcher& /*dispatcher*/) {
+    OptRef<const Network::DownstreamTransportSocketFactory> transport_socket_factory,
+    Envoy::Event::Dispatcher& dispatcher) {
+  if (session->connection()->version().handshake_protocol == quic::PROTOCOL_TLS1_3) {
```
Comment (danzh2010) [inline source/extensions/quic/crypto_stream/envoy_quic_crypto_server_stream.cc:16]: This is always true as Envoy only support RFC and draft29 implementations.

File: source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h
Comment (danzh2010) [inline source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h:1]: I could imagine one day Envoy will need its own QUIC handshaker implementation, but for if only for signing, would overriding EnvoyQuicProofSource::ComputeTlsSignature() serve the purpose?

File: source/extensions/quic/crypto_stream/envoy_quic_crypto_server_stream.cc (hunk 1 above)
Comment (soulxu) [reply source/extensions/quic/crypto_stream/envoy_quic_crypto_server_stream.cc:16]: got it, thanks! I confused when I found some of tests provide other implementation https://github.com/envoyproxy/envoy/blob/372a262894f028431cb84f3a1d361fb8fb6a2518/test/common/quic/envoy_quic_server_session_test.cc#L128-L137

But at same time, I didn't we have any integration test for quic crypto. 

That is probably something we can cleanup

File: source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h
Comment (soulxu) [reply source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h:1]: I remember I tried `EnvoyQuicProofSource::ComputeTlsSignature()` ( I wrote those code months ago, try my best to memory). But it seems the interface of `ComputeTlsSignature` isn't enough. the reasons are:

* There shouldn't be a way to get the raw SSL handle. That is needed by the private key method interface.
* There is no way to invoke the `complete()` method of private key methods.

I guess the real reason is the PrivateKeyProvider's interface is full design to match the boringSSL private key method.

**Comment by ggreenway** [comment]
This looks like it might conflict a bit with #32260 (which is almost finished)

File: source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h
Comment (danzh2010) [reply source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h:1]: I see. The interface is coupled with a SSL object. However, PrivateKeySign() and PrivateKeyComplete() also manages async signing and signature validation. Completely overriding them would require each private key method extension to re-implement these features. 

How about overriding the TlsServerHandshaker::MaybeCreateProofSourceHandle() to provide a handle implementation that calls into sign() and complete() provided by private_method extension. The handle should have access to the SSL object.

Comment (soulxu) [reply source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.h:1]: > 
> How about overriding the TlsServerHandshaker::MaybeCreateProofSourceHandle() to provide a handle implementation that calls into sign() and complete() provided by private_method extension. The handle should have access to the SSL object.

For the SSL object, that will work. But for the complete method, it needs some buffer address parameter which is passed from the boringSSL. That means we need to create a new method for the `ProofSourceHandle`.

In the end, we will have a new ProofSourceHandle interface, and we need a customize TlsServerHandshaker also for invoking that new complete() methid in ProofSourceHandle interface

**Comment by soulxu** [comment]
> This looks like it might conflict a bit with #32260 (which is almost finished)

thanks! I will take a look at.

File: source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+ssl_private_key_result_t EnvoyTlsServerHandshaker::PrivateKeySign(uint8_t* out, size_t* out_len,
+                                                                  size_t max_out, uint16_t sig_alg,
+                                                                  absl::string_view in) {
+  if (transport_socket_factory_->getTlsCertificates().size() > 0) {
+    // TODO(soulxu): Currently the QUIC transport socket only support one certificate. After
+    // QUIC transport socket with multiple certificates, we will figure out how to support
+    // multiple certificates for private key provider also.
+    auto private_key_method =
+        transport_socket_factory_->getTlsCertificates()[0].get().privateKeyMethod();
```
Comment (soulxu) [inline source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.cc:30]: @ggreenway, It seems like I only need to invoke `getTlsCertificateAndKey` method of transport socket factory. And then I only to figure out how to get the sni hostname. I should be able get that from:
https://github.com/google/quiche/blob/02047e04d076d36226e92847a94acf33544f10d6/quiche/quic/core/tls_server_handshaker.h#L100-L101

**Pull request closed without merging**